host/xtest/regression_4000.c

```diff
@@ -5136,3 +5389,848 @@ static void xtest_tee_test_4011(ADBG_Case_t *c)
+{
+	CK_RV rv;
+	CK_SLOT_ID slot;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION;
+	size_t n;
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		return;
+
+	Do_ADBG_BeginSubCase(c, "Check keys against their allowed mechanisms");
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		goto bail;
+
+	for (n = 0; n < ARRAY_SIZE(cktest_allowed_valid); n++) {
+
+		rv = cktest_cipher_init(c, session,
+					cktest_allowed_valid[n].attr_key,
+					cktest_allowed_valid[n].attr_count,
+					cktest_allowed_valid[n].mechanism,
+					TEE_MODE_ENCRYPT,
+					CKR_OK);
+		if (rv)
+			goto bail;
+	}
+
+	for (n = 0; n > ARRAY_SIZE(cktest_allowed_invalid); n++) {
+
+		rv = cktest_cipher_init(c, session,
+					cktest_allowed_valid[n].attr_key,
+					cktest_allowed_valid[n].attr_count,
+					cktest_allowed_valid[n].mechanism,
+					TEE_MODE_ENCRYPT,
+					CKR_KEY_FUNCTION_NOT_PERMITTED);
+		if (rv)
+			goto bail;
+	}
+
+bail:
+	rv = C_CloseSession(session);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK);
+
+	rv = close_lib();
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK);
+
+	Do_ADBG_EndSubCase(c, NULL);
+}
+
+static void xtest_tee_test_4012_allowed_function(ADBG_Case_t *c)
+{
+	CK_RV rv;
+	CK_SLOT_ID slot;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION;
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		return;
+
+	Do_ADBG_BeginSubCase(c, "Check keys against their allowed mechanisms");
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		goto bail;
+
+	rv = cktest_cipher_init(c, session,
+				cktest_aes_enc_only_cts,
+				ARRAY_SIZE(cktest_aes_enc_only_cts),
+				&cktest_aes_cts_mechanism1,
+				TEE_MODE_ENCRYPT,
+				CKR_OK);
+	if (rv)
+		goto bail;
+
+	rv = cktest_cipher_init(c, session,
+				cktest_aes_enc_only_cts,
+				ARRAY_SIZE(cktest_aes_enc_only_cts),
+				&cktest_aes_cts_mechanism1,
+				TEE_MODE_DECRYPT,
+				CKR_KEY_FUNCTION_NOT_PERMITTED);
+	if (rv)
+		goto bail;
+
+	rv = cktest_cipher_init(c, session,
+				cktest_aes_dec_only_ctr,
+				ARRAY_SIZE(cktest_aes_dec_only_ctr),
+				&cktest_aes_ctr_mechanism,
+				TEE_MODE_ENCRYPT,
+				CKR_KEY_FUNCTION_NOT_PERMITTED);
+	if (rv)
+		goto bail;
+
+	rv = cktest_cipher_init(c, session,
+				cktest_aes_dec_only_ctr,
+				ARRAY_SIZE(cktest_aes_dec_only_ctr),
+				&cktest_aes_ctr_mechanism,
+				TEE_MODE_DECRYPT,
+				CKR_OK);
+	if (rv)
+		goto bail;
+
+bail:
+	rv = C_CloseSession(session);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK);
+
+	rv = close_lib();
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK);
+
+	Do_ADBG_EndSubCase(c, NULL);
+}
+
+static void xtest_tee_test_4012(ADBG_Case_t *c)
+{
+	xtest_tee_test_4012_token_info(c);
```

Comment:
The numbers are cheap and we have plenty of them. I think each of these functions deserves its own number and suite entry. This makes it easier when you're only interested in testing a more specific thing, for instance while debugging an error.

We should consider putting all the PKCS#11 test in a 41xx series (and if possible in a separate file, regression_4100.c) since it's a bit different compared to the regular 40xx cases.

Reply:
Fine, I'll move them the **regression_4100.c**.